### ISSUE TITLE
Auth0 Support

### ISSUE THREAD

**Issue by yeager-j** [FoalTS/foal#320]
Auth0 is a complete authentication solution, allowing developers to completely outsource authentication. It allows for email/password as well as OAuth2 and 2fa. 

The problem with Auth0 is it requires a bunch of setup in the backend. Having FoalTS do this automatically as well as providing hooks would be amazing.

Auth0 supports permissions and groups (called 'roles'), so a potential hook could be `@RequireRole('admin')`.

https://auth0.com/docs/quickstart/backend/nodejs

**Comment by nobleach** [comment]
Probably one of the fastest ways to get going with something like this is to take a look at https://github.com/FoalTS/foal/blob/master/packages/jwt/src/jwt.hook.ts One could modify this to support Auth0's authorization/authentication scheme. I'm not sure the responsibility to implement this would fall on the project. (It would be a wonderful contribution from the community)

**Comment by LoicPoullain** [comment]
Hey @yeager-j ,

I'm going to move this card to `Maybe`. Implementing this feature would require a lot of work imo:
- Understand the Auth0 API (usage without Express).
- Design and write all the unit and acceptance tests (there would be a lot as we're dealing with a complex and critical subject).
- Ask for the *Open Source Program* to get Auth0 for free.
- A lot of thought about security concerns.
- A lot of thought about how to integrate the "third-party service" (integration with travis, docker, FoalTS code, etc)

This is probably a useful feature and maybe FoalTS will offer something out-of-the-box one day. For now, as time is limited and I have to choose which features come first, I'll first focus on opensource and maybe more largely used technos such as OpenAPI / Swagger.

**Comment by LoicPoullain** [comment]
@yeager-j I'm going to implement a new feature on the `@JWTRequired` hook to support dynamic secrets. This way, we will be able to dynamically retrieve RSA public keys from a JWKS endpoint and therefore support server-side authentication with Auth0 & AWS Cognito (adding @phptuts).

The implementation will use the `jwks-rsa` package so that we can support *caching* and *rate limiting*.

Here what I have in mind (based on [this example](https://auth0.com/docs/quickstart/backend/nodejs)):

```typescript
import * as jwksClient from 'jwks-rsa';
import { JWTRequired, getRSAPublicKeyFromJWKS } from '@foal/jwt';

const client = jwksClient({
  cache: true,
  rateLimit: true,
  jwksRequestsPerMinute: 5,
  jwksUri: Config.get('jwks.endpoint')
});

@JWTRequired({ secretOrPublicKey: getRSAPublicKeyFromJWKS(client) })
export class ApiController {}
```

If you have any concerns about this, please let me know.

I have a question though, how do you manage authentication on your local host during development? Do you use another Auth0 app? Or do you have a custom solution that does require a connection to Auth0?

Thanks

**Comment by LoicPoullain** [comment]
Feature included in v0.8.9 (released this week or next week).
